controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStoreHelper.java

```diff
@@ -0,0 +1,403 @@
+                    case ConnectionDropped:
+                    case ConnectionFailed:
+                    case UnknownHost:
+                        toThrow = StoreException.create(StoreException.Type.CONNECTION_ERROR, wcfe, errorMessage);
+                        break;
+                    case PreconditionFailed:
+                        toThrow = StoreException.create(StoreException.Type.ILLEGAL_STATE, wcfe, errorMessage);
+                        break;
+                    case AuthFailed:
+                        authToken.set(authHelper.retrieveMasterToken());
+                        toThrow = StoreException.create(StoreException.Type.CONNECTION_ERROR, wcfe, errorMessage);
+                        break;
+                    case SegmentDoesNotExist:
+                        toThrow = StoreException.create(StoreException.Type.DATA_NOT_FOUND, wcfe, errorMessage);
+                        break;
+                    case TableSegmentNotEmpty:
+                        toThrow = StoreException.create(StoreException.Type.DATA_CONTAINS_ELEMENTS, wcfe, errorMessage);
+                        break;
+                    case TableKeyDoesNotExist:
+                        toThrow = StoreException.create(StoreException.Type.DATA_NOT_FOUND, wcfe, errorMessage);
+                        break;
+                    case TableKeyBadVersion:
+                        toThrow = StoreException.create(StoreException.Type.WRITE_CONFLICT, wcfe, errorMessage);
+                        break;
+                    default:
+                        toThrow = StoreException.create(StoreException.Type.UNKNOWN, wcfe, errorMessage);
+                }
+            } else if (cause instanceof HostStoreException) {
+                log.warn("Host Store exception {}", cause.getMessage());
+                toThrow = StoreException.create(StoreException.Type.CONNECTION_ERROR, cause, errorMessage);
+            } else {
+                log.warn("exception of unknown type thrown {} ", errorMessage, cause);
+                toThrow = StoreException.create(StoreException.Type.UNKNOWN, cause, errorMessage);
+            }
+
+            throw new CompletionException(toThrow);
+        });
+    }
+
+    /*
+     * We dont want to do indefinite retries because for controller's graceful shutting down, it waits on grpc service to
+     * be terminated which in turn waits on all outstanding grpc calls to complete. And the store may stall the calls if
+     * there is indefinite retries. Restricting it to 12 retries gives us ~60 seconds worth of wait on the upper side.
+     * Also, note that the call can fail because hostContainerMap has not been updated or it can fail because it cannot
+     * talk to segment store. Both these are translated to ConnectionErrors and are retried. All other exceptions
+     * are thrown back
+     */
+    private <T> CompletableFuture<T> withRetries(Supplier<CompletableFuture<T>> futureSupplier, Supplier<String> errorMessage) {
```

Comment:
If retries might be blocking a controller restart, then perhaps we should have a signaling mechanism that breaks all relevant retry loops once the instance is starting to retry. That would enable us to not wait unnecessarily.

Reply:
That is a good suggestion. I will take it up separately in a follow up PR.

Reply:
I'm glad you understood the comment, one sentence didn't make much sense, I fixed it.